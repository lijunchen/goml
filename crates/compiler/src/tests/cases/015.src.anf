fn print_int_list(xs/0: IntList) -> Unit {
  match xs/0 {
    Tag_0 => {
      string_print("Nil")
    },
    Tag_1 => {
      let x0 = xs/0.constr.1.0 in
      let x1 = xs/0.constr.1.1 in
      let xs/2 = x1 in
      let x/1 = x0 in
      let mtmp2 = string_print("Cons") in
      let mtmp3 = string_print("(") in
      let t25 = int_to_string(x/1) in
      let mtmp4 = string_print(t25) in
      let mtmp5 = string_print(", ") in
      let mtmp6 = print_int_list(xs/2) in
      let mtmp7 = string_print(")") in
      ()
    },
  }
}

fn int_list_rev_aux(xs/3: IntList, acc/4: IntList) -> IntList {
  match xs/3 {
    Tag_0 => {
      acc/4
    },
    Tag_1 => {
      let x8 = xs/3.constr.1.0 in
      let x9 = xs/3.constr.1.1 in
      let tail/6 = x9 in
      let head/5 = x8 in
      let t26 = IntList::Cons(head/5, acc/4) in
      int_list_rev_aux(tail/6, t26)
    },
  }
}

fn int_list_rev(xs/7: IntList) -> IntList {
  let t27 = Tag_0 in
  int_list_rev_aux(xs/7, t27)
}

fn int_list_length(xs/8: IntList) -> Int {
  match xs/8 {
    Tag_0 => {
      0
    },
    Tag_1 => {
      let x10 = xs/8.constr.1.0 in
      let x11 = xs/8.constr.1.1 in
      let xs/9 = x11 in
      let t28 = int_list_length(xs/9) in
      int_add(1, t28)
    },
  }
}

fn print_int_list_length(xs/10: IntList) -> Unit {
  let mtmp12 = string_print("Length: ") in
  let t30 = int_list_length(xs/10) in
  let t29 = int_to_string(t30) in
  let mtmp13 = string_println(t29) in
  ()
}

fn main() -> Unit {
  let x/11 = Tag_0 in
  let mtmp14 = print_int_list(x/11) in
  let mtmp15 = string_println("") in
  let mtmp16 = print_int_list_length(x/11) in
  let t31 = Tag_0 in
  let x/12 = IntList::Cons(1, t31) in
  let mtmp17 = print_int_list(x/12) in
  let mtmp18 = string_println("") in
  let mtmp19 = print_int_list_length(x/12) in
  let t34 = Tag_0 in
  let t33 = IntList::Cons(3, t34) in
  let t32 = IntList::Cons(2, t33) in
  let x/13 = IntList::Cons(1, t32) in
  let mtmp20 = print_int_list(x/13) in
  let mtmp21 = string_println("") in
  let mtmp22 = print_int_list_length(x/13) in
  let y/14 = int_list_rev(x/13) in
  let mtmp23 = print_int_list(y/14) in
  let mtmp24 = string_println("") in
  ()
}