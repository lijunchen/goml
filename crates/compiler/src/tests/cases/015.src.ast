enum IntList {
    Nil,
    Cons(Int, IntList)
}

fn print_int_list(xs:IntList) {
    match xs {
        Nil => print("Nil"),
        Cons(x, xs) => let _ = print("Cons") in
          let _ = print("(") in
          let _ = print(int_to_string(x)) in
          let _ = print(", ") in
          let _ = print_int_list(xs) in
          let _ = print(")") in
          (),
    }
}

fn main() {
    let x = Nil in
    let _ = print_int_list(x) in
    let _ = println("") in
    let x = Cons(1, Nil) in
    let _ = print_int_list(x) in
    let _ = println("") in
    let x = Cons(1, Cons(2, Cons(3, Nil))) in
    let _ = print_int_list(x) in
    let _ = println("") in
    ()
}

