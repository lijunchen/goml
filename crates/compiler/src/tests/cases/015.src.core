fn print_int_list(xs/0: IntList) -> Unit {
  match xs/0 {
    IntList::Nil => {
      print("Nil")
    },
    IntList::Cons(x0, x1) => {
      let x0 = IntList_1_0(xs/0) in
      let x1 = IntList_1_1(xs/0) in
      let xs/2 = x1 in
      let x/1 = x0 in
      let mtmp2 = print("Cons") in
      let mtmp3 = print("(") in
      let mtmp4 = print(int_to_string(x/1)) in
      let mtmp5 = print(", ") in
      let mtmp6 = print_int_list(xs/2) in
      let mtmp7 = print(")") in
      ()
    },
  }
}

fn main() -> Unit {
  let x/3 = IntList::Nil in
  let mtmp8 = print_int_list(x/3) in
  let mtmp9 = println("") in
  let x/4 = IntList::Cons(1, IntList::Nil) in
  let mtmp10 = print_int_list(x/4) in
  let mtmp11 = println("") in
  let x/5 = IntList::Cons(1, IntList::Cons(2, IntList::Cons(3, IntList::Nil))) in
  let mtmp12 = print_int_list(x/5) in
  let mtmp13 = println("") in
  ()
}