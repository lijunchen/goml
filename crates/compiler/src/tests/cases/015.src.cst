FILE@0..701
  ENUM@0..51
    EnumKeyword@0..4 "enum"
    Whitespace@4..5 " "
    Uident@5..12 "IntList"
    Whitespace@12..13 " "
    LBrace@13..14 "{"
    Whitespace@14..19 "\n    "
    VARIANT_LIST@19..51
      VARIANT@19..22
        Uident@19..22 "Nil"
      Comma@22..23 ","
      Whitespace@23..28 "\n    "
      VARIANT@28..46
        Uident@28..32 "Cons"
        TYPE_LIST@32..46
          LParen@32..33 "("
          TYPE_INT@33..36
            IntKeyword@33..36 "Int"
          Comma@36..37 ","
          Whitespace@37..38 " "
          TYPE_ENUM@38..45
            Uident@38..45 "IntList"
          RParen@45..46 ")"
      Comma@46..47 ","
      Whitespace@47..48 "\n"
      RBrace@48..49 "}"
      Whitespace@49..51 "\n\n"
  FN@51..405
    FnKeyword@51..53 "fn"
    Whitespace@53..54 " "
    Lident@54..68 "print_int_list"
    PARAM_LIST@68..82
      LParen@68..69 "("
      PARAM@69..80
        Lident@69..71 "xs"
        Colon@71..72 ":"
        Whitespace@72..73 " "
        TYPE_ENUM@73..80
          Uident@73..80 "IntList"
      RParen@80..81 ")"
      Whitespace@81..82 " "
    BLOCK@82..405
      LBrace@82..83 "{"
      Whitespace@83..88 "\n    "
      EXPR_MATCH@88..402
        MatchKeyword@88..93 "match"
        Whitespace@93..94 " "
        EXPR_LIDENT@94..97
          Lident@94..96 "xs"
          Whitespace@96..97 " "
        MATCH_ARM_LIST@97..402
          LBrace@97..98 "{"
          Whitespace@98..107 "\n        "
          MATCH_ARM@107..126
            PATTERN_CONSTR@107..111
              Uident@107..110 "Nil"
              Whitespace@110..111 " "
            FatArrow@111..113 "=>"
            Whitespace@113..114 " "
            EXPR_CALL@114..126
              EXPR_LIDENT@114..119
                Lident@114..119 "print"
              ARG_LIST@119..126
                LParen@119..120 "("
                ARG@120..125
                  EXPR_STR@120..125
                    Str@120..125 "\"Nil\""
                RParen@125..126 ")"
          Comma@126..127 ","
          Whitespace@127..136 "\n        "
          MATCH_ARM@136..400
            PATTERN_CONSTR@136..148
              Uident@136..140 "Cons"
              LParen@140..141 "("
              PATTERN_VARIABLE@141..142
                Lident@141..142 "x"
              Comma@142..143 ","
              Whitespace@143..144 " "
              PATTERN_VARIABLE@144..146
                Lident@144..146 "xs"
              RParen@146..147 ")"
              Whitespace@147..148 " "
            FatArrow@148..150 "=>"
            Whitespace@150..163 "\n            "
            EXPR_LET@163..400
              LetKeyword@163..166 "let"
              Whitespace@166..167 " "
              PATTERN_WILDCARD@167..169
                WildcardKeyword@167..168 "_"
                Whitespace@168..169 " "
              Eq@169..170 "="
              Whitespace@170..171 " "
              EXPR_LET_VALUE@171..185
                EXPR_CALL@171..185
                  EXPR_LIDENT@171..176
                    Lident@171..176 "print"
                  ARG_LIST@176..185
                    LParen@176..177 "("
                    ARG@177..183
                      EXPR_STR@177..183
                        Str@177..183 "\"Cons\""
                    RParen@183..184 ")"
                    Whitespace@184..185 " "
              InKeyword@185..187 "in"
              Whitespace@187..200 "\n            "
              EXPR_LET_BODY@200..400
                EXPR_LET@200..400
                  LetKeyword@200..203 "let"
                  Whitespace@203..204 " "
                  PATTERN_WILDCARD@204..206
                    WildcardKeyword@204..205 "_"
                    Whitespace@205..206 " "
                  Eq@206..207 "="
                  Whitespace@207..208 " "
                  EXPR_LET_VALUE@208..219
                    EXPR_CALL@208..219
                      EXPR_LIDENT@208..213
                        Lident@208..213 "print"
                      ARG_LIST@213..219
                        LParen@213..214 "("
                        ARG@214..217
                          EXPR_STR@214..217
                            Str@214..217 "\"(\""
                        RParen@217..218 ")"
                        Whitespace@218..219 " "
                  InKeyword@219..221 "in"
                  Whitespace@221..234 "\n            "
                  EXPR_LET_BODY@234..400
                    EXPR_LET@234..400
                      LetKeyword@234..237 "let"
                      Whitespace@237..238 " "
                      PATTERN_WILDCARD@238..240
                        WildcardKeyword@238..239 "_"
                        Whitespace@239..240 " "
                      Eq@240..241 "="
                      Whitespace@241..242 " "
                      EXPR_LET_VALUE@242..266
                        EXPR_CALL@242..266
                          EXPR_LIDENT@242..247
                            Lident@242..247 "print"
                          ARG_LIST@247..266
                            LParen@247..248 "("
                            ARG@248..264
                              EXPR_CALL@248..264
                                EXPR_LIDENT@248..261
                                  Lident@248..261 "int_to_string"
                                ARG_LIST@261..264
                                  LParen@261..262 "("
                                  ARG@262..263
                                    EXPR_LIDENT@262..263
                                      Lident@262..263 "x"
                                  RParen@263..264 ")"
                            RParen@264..265 ")"
                            Whitespace@265..266 " "
                      InKeyword@266..268 "in"
                      Whitespace@268..281 "\n            "
                      EXPR_LET_BODY@281..400
                        EXPR_LET@281..400
                          LetKeyword@281..284 "let"
                          Whitespace@284..285 " "
                          PATTERN_WILDCARD@285..287
                            WildcardKeyword@285..286 "_"
                            Whitespace@286..287 " "
                          Eq@287..288 "="
                          Whitespace@288..289 " "
                          EXPR_LET_VALUE@289..301
                            EXPR_CALL@289..301
                              EXPR_LIDENT@289..294
                                Lident@289..294 "print"
                              ARG_LIST@294..301
                                LParen@294..295 "("
                                ARG@295..299
                                  EXPR_STR@295..299
                                    Str@295..299 "\", \""
                                RParen@299..300 ")"
                                Whitespace@300..301 " "
                          InKeyword@301..303 "in"
                          Whitespace@303..316 "\n            "
                          EXPR_LET_BODY@316..400
                            EXPR_LET@316..400
                              LetKeyword@316..319 "let"
                              Whitespace@319..320 " "
                              PATTERN_WILDCARD@320..322
                                WildcardKeyword@320..321 "_"
                                Whitespace@321..322 " "
                              Eq@322..323 "="
                              Whitespace@323..324 " "
                              EXPR_LET_VALUE@324..344
                                EXPR_CALL@324..344
                                  EXPR_LIDENT@324..338
                                    Lident@324..338 "print_int_list"
                                  ARG_LIST@338..344
                                    LParen@338..339 "("
                                    ARG@339..341
                                      EXPR_LIDENT@339..341
                                        Lident@339..341 "xs"
                                    RParen@341..342 ")"
                                    Whitespace@342..344 "  "
                              InKeyword@344..346 "in"
                              Whitespace@346..359 "\n            "
                              EXPR_LET_BODY@359..400
                                EXPR_LET@359..400
                                  LetKeyword@359..362 "let"
                                  Whitespace@362..363 " "
                                  PATTERN_WILDCARD@363..365
                                    WildcardKeyword@363..364 "_"
                                    Whitespace@364..365 " "
                                  Eq@365..366 "="
                                  Whitespace@366..367 " "
                                  EXPR_LET_VALUE@367..378
                                    EXPR_CALL@367..378
                                      EXPR_LIDENT@367..372
                                        Lident@367..372 "print"
                                      ARG_LIST@372..378
                                        LParen@372..373 "("
                                        ARG@373..376
                                          EXPR_STR@373..376
                                            Str@373..376 "\")\""
                                        RParen@376..377 ")"
                                        Whitespace@377..378 " "
                                  InKeyword@378..380 "in"
                                  Whitespace@380..393 "\n            "
                                  EXPR_LET_BODY@393..400
                                    EXPR_UNIT@393..400
                                      LParen@393..394 "("
                                      RParen@394..395 ")"
                                      Whitespace@395..400 "\n    "
          RBrace@400..401 "}"
          Whitespace@401..402 "\n"
      RBrace@402..403 "}"
      Whitespace@403..405 "\n\n"
  FN@405..701
    FnKeyword@405..407 "fn"
    Whitespace@407..408 " "
    Lident@408..412 "main"
    PARAM_LIST@412..415
      LParen@412..413 "("
      RParen@413..414 ")"
      Whitespace@414..415 " "
    BLOCK@415..701
      LBrace@415..416 "{"
      Whitespace@416..421 "\n    "
      EXPR_LET@421..700
        LetKeyword@421..424 "let"
        Whitespace@424..425 " "
        PATTERN_VARIABLE@425..427
          Lident@425..426 "x"
          Whitespace@426..427 " "
        Eq@427..428 "="
        Whitespace@428..429 " "
        EXPR_LET_VALUE@429..433
          EXPR_UIDENT@429..433
            Uident@429..432 "Nil"
            Whitespace@432..433 " "
        InKeyword@433..435 "in"
        Whitespace@435..440 "\n    "
        EXPR_LET_BODY@440..700
          EXPR_LET@440..700
            LetKeyword@440..443 "let"
            Whitespace@443..444 " "
            PATTERN_WILDCARD@444..446
              WildcardKeyword@444..445 "_"
              Whitespace@445..446 " "
            Eq@446..447 "="
            Whitespace@447..448 " "
            EXPR_LET_VALUE@448..466
              EXPR_CALL@448..466
                EXPR_LIDENT@448..462
                  Lident@448..462 "print_int_list"
                ARG_LIST@462..466
                  LParen@462..463 "("
                  ARG@463..464
                    EXPR_LIDENT@463..464
                      Lident@463..464 "x"
                  RParen@464..465 ")"
                  Whitespace@465..466 " "
            InKeyword@466..468 "in"
            Whitespace@468..473 "\n    "
            EXPR_LET_BODY@473..700
              EXPR_LET@473..700
                LetKeyword@473..476 "let"
                Whitespace@476..477 " "
                PATTERN_WILDCARD@477..479
                  WildcardKeyword@477..478 "_"
                  Whitespace@478..479 " "
                Eq@479..480 "="
                Whitespace@480..481 " "
                EXPR_LET_VALUE@481..493
                  EXPR_CALL@481..493
                    EXPR_LIDENT@481..488
                      Lident@481..488 "println"
                    ARG_LIST@488..493
                      LParen@488..489 "("
                      ARG@489..491
                        EXPR_STR@489..491
                          Str@489..491 "\"\""
                      RParen@491..492 ")"
                      Whitespace@492..493 " "
                InKeyword@493..495 "in"
                Whitespace@495..501 "\n\n    "
                EXPR_LET_BODY@501..700
                  EXPR_LET@501..700
                    LetKeyword@501..504 "let"
                    Whitespace@504..505 " "
                    PATTERN_VARIABLE@505..507
                      Lident@505..506 "x"
                      Whitespace@506..507 " "
                    Eq@507..508 "="
                    Whitespace@508..509 " "
                    EXPR_LET_VALUE@509..522
                      EXPR_CALL@509..522
                        EXPR_UIDENT@509..513
                          Uident@509..513 "Cons"
                        ARG_LIST@513..522
                          LParen@513..514 "("
                          ARG@514..517
                            EXPR_INT@514..515
                              Int32@514..515 "1"
                            Comma@515..516 ","
                            Whitespace@516..517 " "
                          ARG@517..520
                            EXPR_UIDENT@517..520
                              Uident@517..520 "Nil"
                          RParen@520..521 ")"
                          Whitespace@521..522 " "
                    InKeyword@522..524 "in"
                    Whitespace@524..529 "\n    "
                    EXPR_LET_BODY@529..700
                      EXPR_LET@529..700
                        LetKeyword@529..532 "let"
                        Whitespace@532..533 " "
                        PATTERN_WILDCARD@533..535
                          WildcardKeyword@533..534 "_"
                          Whitespace@534..535 " "
                        Eq@535..536 "="
                        Whitespace@536..537 " "
                        EXPR_LET_VALUE@537..555
                          EXPR_CALL@537..555
                            EXPR_LIDENT@537..551
                              Lident@537..551 "print_int_list"
                            ARG_LIST@551..555
                              LParen@551..552 "("
                              ARG@552..553
                                EXPR_LIDENT@552..553
                                  Lident@552..553 "x"
                              RParen@553..554 ")"
                              Whitespace@554..555 " "
                        InKeyword@555..557 "in"
                        Whitespace@557..562 "\n    "
                        EXPR_LET_BODY@562..700
                          EXPR_LET@562..700
                            LetKeyword@562..565 "let"
                            Whitespace@565..566 " "
                            PATTERN_WILDCARD@566..568
                              WildcardKeyword@566..567 "_"
                              Whitespace@567..568 " "
                            Eq@568..569 "="
                            Whitespace@569..570 " "
                            EXPR_LET_VALUE@570..582
                              EXPR_CALL@570..582
                                EXPR_LIDENT@570..577
                                  Lident@570..577 "println"
                                ARG_LIST@577..582
                                  LParen@577..578 "("
                                  ARG@578..580
                                    EXPR_STR@578..580
                                      Str@578..580 "\"\""
                                  RParen@580..581 ")"
                                  Whitespace@581..582 " "
                            InKeyword@582..584 "in"
                            Whitespace@584..590 "\n\n    "
                            EXPR_LET_BODY@590..700
                              EXPR_LET@590..700
                                LetKeyword@590..593 "let"
                                Whitespace@593..594 " "
                                PATTERN_VARIABLE@594..596
                                  Lident@594..595 "x"
                                  Whitespace@595..596 " "
                                Eq@596..597 "="
                                Whitespace@597..598 " "
                                EXPR_LET_VALUE@598..629
                                  EXPR_CALL@598..629
                                    EXPR_UIDENT@598..602
                                      Uident@598..602 "Cons"
                                    ARG_LIST@602..629
                                      LParen@602..603 "("
                                      ARG@603..606
                                        EXPR_INT@603..604
                                          Int32@603..604 "1"
                                        Comma@604..605 ","
                                        Whitespace@605..606 " "
                                      ARG@606..627
                                        EXPR_CALL@606..627
                                          EXPR_UIDENT@606..610
                                            Uident@606..610 "Cons"
                                          ARG_LIST@610..627
                                            LParen@610..611 "("
                                            ARG@611..614
                                              EXPR_INT@611..612
                                                Int32@611..612 "2"
                                              Comma@612..613 ","
                                              Whitespace@613..614 " "
                                            ARG@614..626
                                              EXPR_CALL@614..626
                                                EXPR_UIDENT@614..618
                                                  Uident@614..618 "Cons"
                                                ARG_LIST@618..626
                                                  LParen@618..619 "("
                                                  ARG@619..622
                                                    EXPR_INT@619..620
                                                      Int32@619..620 "3"
                                                    Comma@620..621 ","
                                                    Whitespace@621..622 " "
                                                  ARG@622..625
                                                    EXPR_UIDENT@622..625
                                                      Uident@622..625 "Nil"
                                                  RParen@625..626 ")"
                                            RParen@626..627 ")"
                                      RParen@627..628 ")"
                                      Whitespace@628..629 " "
                                InKeyword@629..631 "in"
                                Whitespace@631..636 "\n    "
                                EXPR_LET_BODY@636..700
                                  EXPR_LET@636..700
                                    LetKeyword@636..639 "let"
                                    Whitespace@639..640 " "
                                    PATTERN_WILDCARD@640..642
                                      WildcardKeyword@640..641 "_"
                                      Whitespace@641..642 " "
                                    Eq@642..643 "="
                                    Whitespace@643..644 " "
                                    EXPR_LET_VALUE@644..662
                                      EXPR_CALL@644..662
                                        EXPR_LIDENT@644..658
                                          Lident@644..658 "print_int_list"
                                        ARG_LIST@658..662
                                          LParen@658..659 "("
                                          ARG@659..660
                                            EXPR_LIDENT@659..660
                                              Lident@659..660 "x"
                                          RParen@660..661 ")"
                                          Whitespace@661..662 " "
                                    InKeyword@662..664 "in"
                                    Whitespace@664..669 "\n    "
                                    EXPR_LET_BODY@669..700
                                      EXPR_LET@669..700
                                        LetKeyword@669..672 "let"
                                        Whitespace@672..673 " "
                                        PATTERN_WILDCARD@673..675
                                          WildcardKeyword@673..674 "_"
                                          Whitespace@674..675 " "
                                        Eq@675..676 "="
                                        Whitespace@676..677 " "
                                        EXPR_LET_VALUE@677..689
                                          EXPR_CALL@677..689
                                            EXPR_LIDENT@677..684
                                              Lident@677..684 "println"
                                            ARG_LIST@684..689
                                              LParen@684..685 "("
                                              ARG@685..687
                                                EXPR_STR@685..687
                                                  Str@685..687 "\"\""
                                              RParen@687..688 ")"
                                              Whitespace@688..689 " "
                                        InKeyword@689..691 "in"
                                        Whitespace@691..697 "\n\n    "
                                        EXPR_LET_BODY@697..700
                                          EXPR_UNIT@697..700
                                            LParen@697..698 "("
                                            RParen@698..699 ")"
                                            Whitespace@699..700 "\n"
      RBrace@700..701 "}"