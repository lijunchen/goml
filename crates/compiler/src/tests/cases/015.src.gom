package main

import (
    "fmt"
)

func unit_to_string(x struct{}) string {
    return "()"
}

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_sub(x int, y int) int {
    return x - y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_print(s string) struct{} {
    fmt.Print(s)
    return struct{}{}
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func missing(s string) struct{} {
    panic("missing: " + s)
    return struct{}{}
}

type IntList interface {
    isIntList()
}

type Nil struct {}

func (_ Nil) isIntList() {}

type Cons struct {
    _0 int
    _1 IntList
}

func (_ Cons) isIntList() {}

func print_int_list(xs__0 IntList) struct{} {
    var ret35 struct{}
    switch xs__0.(type) {
    case Nil:
        ret35 = string_print("Nil")
    case Cons:
        var x0 int = xs__0.(Cons)._0
        var x1 IntList = xs__0.(Cons)._1
        var xs__2 IntList = x1
        var x__1 int = x0
        var mtmp2 struct{} = string_print("Cons")
        var mtmp3 struct{} = string_print("(")
        var t25 string = int_to_string(x__1)
        var mtmp4 struct{} = string_print(t25)
        var mtmp5 struct{} = string_print(", ")
        var mtmp6 struct{} = print_int_list(xs__2)
        var mtmp7 struct{} = string_print(")")
        ret35 = struct{}{}
    }
    return ret35
}

func int_list_rev_aux(xs__3 IntList, acc__4 IntList) IntList {
    var ret36 IntList
    switch xs__3.(type) {
    case Nil:
        ret36 = acc__4
    case Cons:
        var x8 int = xs__3.(Cons)._0
        var x9 IntList = xs__3.(Cons)._1
        var tail__6 IntList = x9
        var head__5 int = x8
        var t26 IntList = Cons{
            _0: head__5,
            _1: acc__4,
        }
        ret36 = int_list_rev_aux(tail__6, t26)
    }
    return ret36
}

func int_list_rev(xs__7 IntList) IntList {
    var ret37 IntList
    var t27 IntList = Nil{}
    ret37 = int_list_rev_aux(xs__7, t27)
    return ret37
}

func int_list_length(xs__8 IntList) int {
    var ret38 int
    switch xs__8.(type) {
    case Nil:
        ret38 = 0
    case Cons:
        var x10 int = xs__8.(Cons)._0
        var x11 IntList = xs__8.(Cons)._1
        var xs__9 IntList = x11
        var t28 int = int_list_length(xs__9)
        ret38 = int_add(1, t28)
    }
    return ret38
}

func print_int_list_length(xs__10 IntList) struct{} {
    var ret39 struct{}
    var mtmp12 struct{} = string_print("Length: ")
    var t30 int = int_list_length(xs__10)
    var t29 string = int_to_string(t30)
    var mtmp13 struct{} = string_println(t29)
    ret39 = struct{}{}
    return ret39
}

func main0() struct{} {
    var ret40 struct{}
    var x__11 IntList = Nil{}
    var mtmp14 struct{} = print_int_list(x__11)
    var mtmp15 struct{} = string_println("")
    var mtmp16 struct{} = print_int_list_length(x__11)
    var t31 IntList = Nil{}
    var x__12 IntList = Cons{
        _0: 1,
        _1: t31,
    }
    var mtmp17 struct{} = print_int_list(x__12)
    var mtmp18 struct{} = string_println("")
    var mtmp19 struct{} = print_int_list_length(x__12)
    var t34 IntList = Nil{}
    var t33 IntList = Cons{
        _0: 3,
        _1: t34,
    }
    var t32 IntList = Cons{
        _0: 2,
        _1: t33,
    }
    var x__13 IntList = Cons{
        _0: 1,
        _1: t32,
    }
    var mtmp20 struct{} = print_int_list(x__13)
    var mtmp21 struct{} = string_println("")
    var mtmp22 struct{} = print_int_list_length(x__13)
    var y__14 IntList = int_list_rev(x__13)
    var mtmp23 struct{} = print_int_list(y__14)
    var mtmp24 struct{} = string_println("")
    ret40 = struct{}{}
    return ret40
}

func main() {
    main0()
}
