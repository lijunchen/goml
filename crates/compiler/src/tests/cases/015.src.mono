fn print_int_list(xs/0: IntList) -> unit {
  match xs/0 {
    IntList::Nil => {
      string_print("Nil")
    },
    IntList::Cons(x0, x1) => {
      let x0 = IntList::Cons._0(xs/0) in
      let x1 = IntList::Cons._1(xs/0) in
      let xs/2 = x1 in
      let x/1 = x0 in
      let mtmp2 = string_print("Cons") in
      let mtmp3 = string_print("(") in
      let mtmp4 = string_print(int_to_string(x/1)) in
      let mtmp5 = string_print(", ") in
      let mtmp6 = print_int_list(xs/2) in
      let mtmp7 = string_print(")") in
      ()
    },
  }
}

fn int_list_rev_aux(xs/3: IntList, acc/4: IntList) -> IntList {
  match xs/3 {
    IntList::Nil => {
      acc/4
    },
    IntList::Cons(x8, x9) => {
      let x8 = IntList::Cons._0(xs/3) in
      let x9 = IntList::Cons._1(xs/3) in
      let tail/6 = x9 in
      let head/5 = x8 in
      int_list_rev_aux(tail/6, IntList::Cons(head/5, acc/4))
    },
  }
}

fn int_list_rev(xs/7: IntList) -> IntList {
  int_list_rev_aux(xs/7, IntList::Nil)
}

fn int_list_length(xs/8: IntList) -> int {
  match xs/8 {
    IntList::Nil => {
      0
    },
    IntList::Cons(x10, x11) => {
      let x10 = IntList::Cons._0(xs/8) in
      let x11 = IntList::Cons._1(xs/8) in
      let xs/9 = x11 in
      int_add(1, int_list_length(xs/9))
    },
  }
}

fn print_int_list_length(xs/10: IntList) -> unit {
  let mtmp12 = string_print("Length: ") in
  let mtmp13 = string_println(int_to_string(int_list_length(xs/10))) in
  ()
}

fn main() -> unit {
  let x/11 = IntList::Nil in
  let mtmp14 = print_int_list(x/11) in
  let mtmp15 = string_println("") in
  let mtmp16 = print_int_list_length(x/11) in
  let x/12 = IntList::Cons(1, IntList::Nil) in
  let mtmp17 = print_int_list(x/12) in
  let mtmp18 = string_println("") in
  let mtmp19 = print_int_list_length(x/12) in
  let x/13 = IntList::Cons(1, IntList::Cons(2, IntList::Cons(3, IntList::Nil))) in
  let mtmp20 = print_int_list(x/13) in
  let mtmp21 = string_println("") in
  let mtmp22 = print_int_list_length(x/13) in
  let y/14 = int_list_rev(x/13) in
  let mtmp23 = print_int_list(y/14) in
  let mtmp24 = string_println("") in
  ()
}