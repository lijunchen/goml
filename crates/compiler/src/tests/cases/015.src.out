()
====
Nil
Cons(1, Nil)
Cons(1, Cons(2, Cons(3, Nil)))
